src/Content/Text/BBCode.php

```diff
@@ -2104,63 +2092,77 @@ public static function getTags($string)
 	{
 		$ret = [];
 
-		// Convert hashtag links to hashtags
-		$string = preg_replace('/#\[url\=([^\[\]]*)\](.*?)\[\/url\]/ism', '#$2 ', $string);
-
-		// ignore anything in a code block
-		$string = preg_replace('/\[code.*?\].*?\[\/code\]/sm', '', $string);
+		BBCode::performWithEscapedTags($string, ['noparse', 'pre', 'code'], function ($string) use (&$ret) {
+			// Convert hashtag links to hashtags
+			$string = preg_replace('/#\[url\=([^\[\]]*)\](.*?)\[\/url\]/ism', '#$2 ', $string);
 
-		// Force line feeds at bbtags
-		$string = str_replace(['[', ']'], ["\n[", "]\n"], $string);
+			// Force line feeds at bbtags
+			$string = str_replace(['[', ']'], ["\n[", "]\n"], $string);
 
-		// ignore anything in a bbtag
-		$string = preg_replace('/\[(.*?)\]/sm', '', $string);
+			// ignore anything in a bbtag
+			$string = preg_replace('/\[(.*?)\]/sm', '', $string);
 
-		// Match full names against @tags including the space between first and last
-		// We will look these up afterward to see if they are full names or not recognisable.
+			// Match full names against @tags including the space between first and last
+			// We will look these up afterward to see if they are full names or not recognisable.
 
-		if (preg_match_all('/(@[^ \x0D\x0A,:?]+ [^ \x0D\x0A@,:?]+)([ \x0D\x0A@,:?]|$)/', $string, $matches)) {
-			foreach ($matches[1] as $match) {
-				if (strstr($match, ']')) {
-					// we might be inside a bbcode color tag - leave it alone
-					continue;
-				}
+			if (preg_match_all('/(@[^ \x0D\x0A,:?]+ [^ \x0D\x0A@,:?]+)([ \x0D\x0A@,:?]|$)/', $string, $matches)) {
+				foreach ($matches[1] as $match) {
+					if (strstr($match, ']')) {
+						// we might be inside a bbcode color tag - leave it alone
+						continue;
+					}
 
-				if (substr($match, -1, 1) === '.') {
-					$ret[] = substr($match, 0, -1);
-				} else {
-					$ret[] = $match;
+					if (substr($match, -1, 1) === '.') {
+						$ret[] = substr($match, 0, -1);
+					} else {
+						$ret[] = $match;
+					}
 				}
 			}
-		}
 
-		// Otherwise pull out single word tags. These can be @nickname, @first_last
-		// and #hash tags.
+			// Otherwise pull out single word tags. These can be @nickname, @first_last
+			// and #hash tags.
 
-		if (preg_match_all('/([!#@][^\^ \x0D\x0A,;:?]+)([ \x0D\x0A,;:?]|$)/', $string, $matches)) {
-			foreach ($matches[1] as $match) {
-				if (strstr($match, ']')) {
-					// we might be inside a bbcode color tag - leave it alone
-					continue;
-				}
+			if (preg_match_all('/([!#@][^\^ \x0D\x0A,;:?\']*[^\^ \x0D\x0A,;:?!\'.])/', $string, $matches)) {
+				foreach ($matches[1] as $match) {
+					if (strstr($match, ']')) {
+						// we might be inside a bbcode color tag - leave it alone
+						continue;
+					}
 
-				if (substr($match, -1, 1) === '.') {
```

Comment:
What was this check about?

Reply:
The previous regular expression was too greedy and could include an ending period in a tag. The new one ensures that the ending character of the match isn't any of ` \x0D\x0A,;:?!'.`.

Reply:
Okay.

BTW: I never will understand this RegExp stuff ...